New: semi-style rule (fixes #8169)

mysticatea opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))

Fixes #8169.

**What changes did you make? (Give an overview)**

This PR adds new rule `semi-style` which enforces locations.

**Is there anything you'd like reviewers to focus on?**

Please check whether the tests and the document are enough.

gyandeeps commented on lib/rules/semi-style.js:
```diff
@@ -0,0 +1,131 @@
+         * @param {Token} right The right token to check.
+         * @returns {boolean} `true` if comments exist between the given 2 tokens.
+         */
+        function commentsExistBetween(left, right) {
```
This function gets created every time rule context is created. Why not just move this function outside so that single declaration can used across different context as this function is pure (or can be made pure).
I know we do this in almost all rules but I think we should change this style.
thoughts?

mysticatea replied:
Good question.
In fact, there is [here](https://github.com/eslint/eslint/pull/8101/files#diff-8129190ebfd8121cbcc8e96f0d25ae80R575) but it has not been merged. I will replace those after that is merged.

gyandeeps commented on docs/rules/semi-style.md:
````diff
@@ -0,0 +1,100 @@
+}
+```
+
+If the option is a string, it means both `statements` and `forLoopHead` are set the same value.
````
Based on other rules, we should list the rule options up top and then give examples for each option.

platinumazure commented on lib/rules/semi-style.js:
```diff
@@ -0,0 +1,131 @@
+                    data: {
+                        pos: (expected === "last")
+                            ? "the last of the previous line"
+                            : "the head of the next line"
```
I think these will read better as "the end of the previous line" and "the beginning of the next line".

mysticatea commented:
Thank you for the review. I'm sorry that my response was slow since my sick.

I updated the document and the error message.

not-an-aardvark commented on lib/rules/semi-style.js:
```diff
@@ -0,0 +1,131 @@
+        "EmptyStatement", "ExportAllDeclaration", "ExportDefaultDeclaration",
+        "ExportNamedDeclaration", "ExpressionStatement", "ImportDeclaration",
+        "ReturnStatement", "ThrowStatement", "VariableDeclaration"
+    ].join(",")
```
Could this be the selector `:statement, :declaration` instead?

mysticatea replied:
I was using `:statement` at first but I switched it to this because I need to ignore `IfStatement`, `ForStatement`, `ForInStatement`, `ForOfStatement` and `WhileStatement` in order to avoid checking in duplicate.

Is it better if I use `:statement` and `if` statement in the handler?

not-an-aardvark commented on docs/rules/semi-style.md:
````diff
@@ -0,0 +1,105 @@
+```
+
+- `statements` ... configuration of statements.
+- `forLoopHead` ... configuration of `for(;;){}`.
````
Is it common for people to have a different preference for `for` loops than they would have for statements? I'm wondering if it's worth having a separate option for loops.

mysticatea replied:
Sure, it's probably OK that this rule always behaves as `"last"` for `for` loops even if the option is `"first"`.

not-an-aardvark commented on docs/rules/semi-style.md:
````diff
@@ -0,0 +1,105 @@
+```js
+/*eslint semi-style: ["error", "last"]*/
+
+var a = 0
````
Why is `var a = 0` incorrect here? There is no semicolon after that statement.

not-an-aardvark commented on docs/rules/semi-style.md:
```diff
@@ -0,0 +1,105 @@
+    ; ++i
+) {
+    foo()
+}
```
If we decide to keep the `forLoopHead` option, I think it would be a good idea to add some examples where `statements` and `forLoopHead` are different.

not-an-aardvark commented on lib/rules/semi-style.js:
```diff
@@ -0,0 +1,131 @@
+        /**
+         * Check the given semicolon token.
+         * @param {Token} semiToken The semicolon token to check.
+         * @param {string} expected The expected location to check.
```
Nitpick: Can you clarify in the comment that `expected` is either `"first"` or `"last"`?

not-an-aardvark commented on lib/rules/semi-style.js:
```diff
@@ -0,0 +1,131 @@
+    create(context) {
+        const sourceCode = context.getSourceCode();
+        const options = context.options[0] || "last";
+        const statementLoc = typeof options === "string" ? options : (options.statements || "last");
```
Nitpick: Can you rename this variables to something like `statementOption`? When I saw `loc` later on in the file, I thought it was referring to a `{ line: number, column: number }` object.

not-an-aardvark commented on docs/rules/semi-style.md:
```diff
@@ -0,0 +1,105 @@
+
+## When Not To Use It
+
+If you don't want to notify the location of semicolons, then it's safe to disable this rule.
```
Can you add a link to the `semi` rule somewhere in this file? If someone is unfamiliar with ESLint and they want to enforce that all lines have semicolons, they might mistakenly think that a rule called `semi-style` would enforce this, so it would be useful to direct them to the `semi` rule.

not-an-aardvark commented on lib/rules/semi-style.js:
```diff
@@ -0,0 +1,131 @@
+
+            ForStatement(node) {
+                const firstSemi = node.init && sourceCode.getTokenAfter(node.init);
+                const secondSemi = node.test && sourceCode.getTokenAfter(node.test);
```
I think this will be incorrect if `node.init` or `node.test` is parenthesized. The following code should be valid, but the rule reports an error:

```js
/* eslint semi-style: [error, "first"] */
for (
  (i = 0)
  ; i < 10
  ; i++
) {}
```

mysticatea commented:
I updated this PR to follow suggestions.

- Fixed 2 bugs (ab8a1ca, 3bcf09c)
- Simplified options; only `"last"` or `"first"`. Semicolons of `for` loop heads should be at the end of lines always.
